lib/ids.js

```diff
@@ -7,36 +11,114 @@ two numbers results in a different checksum digit, and thus an invalid id.
-  var digitArray = digits.split('');
-
-  var offset = digitArray.length + 1;
-  var total = _.reduce(digitArray, function(sum, digit, index) {
-    return sum + (Number(digit) * (offset - index));
-  }, 0);
-  var result = total % 11;
+const addCheckDigit = (digits) => {
+  const digitArray = digits.split('');
+
+  const offset = digitArray.length + 1;
+  const total = _.reduce(
+    digitArray,
+    (sum, digit, index) => sum + (Number(digit) * (offset - index)),
+    0
+  );
+
+  const result = total % 11;
   digitArray.push(result === 10 ? 0 : result);
 
   return digitArray.join('');
 };
 
-module.exports = {
-    /*
-      Generates a random N digit ID. The last ID is a checksum digit
-    */
-    generate: function(length) {
-      if (length && typeof length !== 'number') {
-        throw new Error('generate requires that you pass it a length 0 < x < 14');
-      }
+const generateId = function(length) {
+  if (length && typeof length !== 'number') {
+    throw new Error('generate requires that you pass it a length 0 < x < 14');
+  }
 
-      if (length >= 14) {
-        // TODO: when we support a million billion patients per installation change
-        //       the algorithm to support it ;-)
-        throw new Error('WARNING: id length of ' + length + ' is too long');
+  if (length >= 14) {
+    // TODO: when we support a million billion patients per installation change
+    //       the algorithm to support it ;-)
+    throw new Error('WARNING: id length of ' + length + ' is too long');
+  }
+
+  const randomDigits = String(Math.random() * 10).replace('.','');
+
+  return addCheckDigit(randomDigits.substring(0, length - 1));
+};
+
+/*
+ * Given a collection of ids return an array of those not used already
+ */
+const findUnusedIds = (db, freshIds) => {
+  return new Promise((resolve, reject) => {
+    db.medic.view('medic', 'registered_patients', {
+      keys: [...freshIds]
+    }, (err, results) => {
+      if (err) {
+        return reject(err);
       }
 
-      var randomDigits = String(Math.random() * 10).replace('.','');
+      const uniqueIds = new Set(freshIds);
+
+      _.pluck(results.rows, 'key').forEach(patientId => {
+        uniqueIds.delete(patientId);
+      });
+
+      resolve(uniqueIds);
+    });
+  });
+};
+
+const generateNewIds = (currentIdLength) => {
+  const freshIds = new Set();
+  do {
+    freshIds.add(generateId(currentIdLength));
+  } while (freshIds.size < MAX_IDS_TO_CACHE);
+
+  return freshIds;
+};
+
+const generator = function*(db) {
+  let cachedIds = new Set().values(),
+      currentIdLength = DEFAULT_ID_LENGTH;
+
+  // Developers NB: if you set the cache size too high it will take forever
+  // or potentially be impossible to actually generate enough unique randomly
+  // generated ids.
+  if (MAX_IDS_TO_CACHE * 10 > Math.pow(10, DEFAULT_ID_LENGTH)) {
+    throw new Error('MAX_IDS_TO_CACHE too high compared to DEFAULT_ID_LENGTH');
+  }
+
+  const getNextValue = function(resolve, reject) {
```

Comment:
You're totally right, I hadn't mentally transitioned back to promises properly. Fixed.